[JIT] Add autocasting to freezing pass & enable autocast pass by default

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #74178

Autocasting + freezing should reduce model size in some scenarios, since half-precision constants should be smaller than full-precision constants. This also enables the jit autocast pass by default, so `torch._C._jit_set_autocast_mode(True)` doesn't need to be set in order to enable autocasting.

Differential Revision: [D34914245](https://our.internmc.facebook.com/intern/diff/D34914245)

davidberard98 committed:
[JIT] Add autocasting to freezing pass & enable autocast pass by default

Autocasting + freezing should reduce model size in some scenarios, since half-precision constants should be smaller than full-precision constants. This also enables the jit autocast pass by default, so `torch._C._jit_set_autocast_mode(True)` doesn't need to be set in order to enable autocasting.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Add autocasting to freezing pass & enable autocast pass by default"


Autocasting + freezing should reduce model size in some scenarios, since half-precision constants should be smaller than full-precision constants. This also enables the jit autocast pass by default, so `torch._C._jit_set_autocast_mode(True)` doesn't need to be set in order to enable autocasting.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Add autocasting to freezing pass & enable autocast pass by default"


Autocasting + freezing should reduce model size in some scenarios, since half-precision constants should be smaller than full-precision constants. This also enables the jit autocast pass by default, so `torch._C._jit_set_autocast_mode(True)` doesn't need to be set in order to enable autocasting.

[ghstack-poisoned]

eellison reviewed (approved):
Nice!

jjsjann123 commented on torch/csrc/jit/passes/autocast.cpp:
```diff
@@ -18,9 +18,7 @@ namespace jit {
-// TODO: Turn on autocast by default. default turned off to avoid tests failures
-// as we prototype the support
-bool autocast_enabled = false;
+bool autocast_enabled = true;
```
:crossed_fingers: Hopefully this would go smoothly

davidberard98 committed:
Update on "[JIT] Add autocasting to freezing pass & enable autocast pass by default"


Autocasting + freezing should reduce model size in some scenarios, since half-precision constants should be smaller than full-precision constants. This also enables the jit autocast pass by default, so `torch._C._jit_set_autocast_mode(True)` doesn't need to be set in order to enable autocasting.

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] Add autocasting to freezing pass & enable autocast pass by default"


Autocasting + freezing should reduce model size in some scenarios, since half-precision constants should be smaller than full-precision constants. This also enables the jit autocast pass by default, so `torch._C._jit_set_autocast_mode(True)` doesn't need to be set in order to enable autocasting.

[ghstack-poisoned]

davidberard98 commented on torch/csrc/jit/passes/autocast.cpp:
```diff
@@ -437,7 +444,9 @@ void handleBlock(Block* block, AutocastContext initial_state) {
       // Banned in autocast, see binary_cross_entropy_banned()
       case aten::binary_cross_entropy:
-        AT_ERROR("Unsafe to autocast");
+        if (has_seen_amp_section) {
+          AT_ERROR("Unsafe to autocast");
+        }
```
is this a reasonable approach for handling binary_cross_entropy?

davidberard98 replied:
or should we be checking `current_state()` to match eager autocasting mode behavior?

jjsjann123 replied:
This is already better than what we had :stuck_out_tongue: 

But it feels a little strange to me. I thought it's easier to just peek `current_state()` and assert if we don't want user to shoot at their own feet. Any specific reason that we want to assert when `binary_cross_entropy` is safely outside the amp enabled context?

davidberard98 replied:
yeah, probably `current_state()` is fine and less likely to break anything that's already running, I'll go change that.

davidberard98 commented on torch/csrc/jit/passes/autocast.cpp:
```diff
@@ -149,17 +148,23 @@ void castTensorInputs(
 
   std::unordered_set<Value*> casted_inputs;
+  // need to also keep the inputs in order, otherwise tracing fails
+  // sanity checks because casting ops are inserted in random order
+  std::vector<Value*> casted_inputs_ordered;
   for (auto input : node->inputs()) {
     // TODO: update cast_op signature to take dynamic context flags
     auto input_tensor_type = input->type()->cast<TensorType>();
     if (input_tensor_type && input->node()->kind() != cast_op) {
-      casted_inputs.insert(input);
+      auto has_inserted = casted_inputs.insert(input);
+      if (has_inserted.second) {
+        casted_inputs_ordered.push_back(input);
+      }
```
fyi added this to fix test failures on [test_nhwc_autocast_jit_trace_model](https://github.com/pytorch/pytorch/blob/master/test/test_jit.py#L16041-L16083)

jjsjann123 replied:
Nice catch! Sorry about the initial failures.

davidberard98 reviewed (commented):
@eellison @jjsjann123 could you take another look now that I fixed a few failing tests?

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34914245).

jjsjann123 reviewed (commented):
minor comment on handling of `binary_cross_entropy`

davidberard98 committed:
Update on "[JIT] Add autocasting to freezing pass & enable autocast pass by default"


Autocasting + freezing should reduce model size in some scenarios, since half-precision constants should be smaller than full-precision constants. This also enables the jit autocast pass by default, so `torch._C._jit_set_autocast_mode(True)` doesn't need to be set in order to enable autocasting.

Differential Revision: [D34914245](https://our.internmc.facebook.com/intern/diff/D34914245)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34914245).

davidberard98 committed:
Update on "[JIT] Add autocasting to freezing pass & enable autocast pass by default"


Autocasting + freezing should reduce model size in some scenarios, since half-precision constants should be smaller than full-precision constants. This also enables the jit autocast pass by default, so `torch._C._jit_set_autocast_mode(True)` doesn't need to be set in order to enable autocasting.

Differential Revision: [D34914245](https://our.internmc.facebook.com/intern/diff/D34914245)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34914245).

davidberard98 committed:
Update on "[JIT] Add autocasting to freezing pass & enable autocast pass by default"


Autocasting + freezing should reduce model size in some scenarios, since half-precision constants should be smaller than full-precision constants. This also enables the jit autocast pass by default, so `torch._C._jit_set_autocast_mode(True)` doesn't need to be set in order to enable autocasting.

Differential Revision: [D34914245](https://our.internmc.facebook.com/intern/diff/D34914245)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34914245).

davidberard98 committed:
Update on "[JIT] Add autocasting to freezing pass & enable autocast pass by default"


Autocasting + freezing should reduce model size in some scenarios, since half-precision constants should be smaller than full-precision constants. This also enables the jit autocast pass by default, so `torch._C._jit_set_autocast_mode(True)` doesn't need to be set in order to enable autocasting.

Differential Revision: [D34914245](https://our.internmc.facebook.com/intern/diff/D34914245)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34914245).

eellison commented on torch/csrc/jit/passes/freeze_module.cpp:
```diff
@@ -101,6 +102,9 @@ class AttributePropagator {
     };
     auto applyOptimizations = [](std::shared_ptr<Graph>& subgraph) {
+#ifndef C10_MOBILE
+      Autocast(subgraph);
```
cc @dhruvbird do you know why mobile is including freezing ? I would think that freezing happens on the server first before mobile build

dhruvbird replied:
The [optimize_for_mobile API](https://pytorch.org/docs/stable/mobile_optimizer.html) basically does a freeze pass since the lite-interpreter doesn't (or didn't till very recently?) support the `CALL` instruction, so we had to inline everything by freezing. Also I think some optimization passes will work better with a frozen module since it could lead to better operator fusion.

cc @iseeyuan @dreiss to confirm.

dreiss replied:
I'm not sure.  optimize_for_mobile runs on the server build.  Our internal production and external recommended mobile settings don't include the JIT at all.

dhruvbird replied:
ah! @dreiss good catch - yes the freeze pass does happen on the server build (and not the mobile build), and `C10_MOBILE` is defined only for the mobile build (and not the server build).

dhruvbird replied:
In either case, this doesn't seem wrong, since this code will only be used when `C10_MOBILE` is not defined, but it does look confusing.

eellison replied:
@dhruvbird do you mind submitting a PR to clean it up?

eellison reviewed (approved):
Nice, LGTM !!
